### ISSUE TITLE
Ability to use relatedQuery() outside of subqueries

### ISSUE THREAD

**Issue by devinivy** [Vincit/objection.js#1365]
Currently in objection v1 `relatedQuery()` can only be used to build subqueries.  In #1302 and on gitter there has been some chatter about making `relatedQuery()` meaningful outside of subqueries.  I think the idea is easiest to illustrate through examples.

```js
// All these queries below are executable but are lazy, i.e. not executed until awaited.
// The new API is called per(), which specifies the base query made against a relation.

/* select * from users where type = 'musician' */
const musicians = User.query().where('type', 'musician');

/* select * from animals where
     type = 'dog' and
     "ownerId" in (select "ownerId" from users where type = 'musician') */
const musiciansDogs = User.relatedQuery('pets').per(musicians).where('type', 'dog');

/* update animals set "ownerId" = null where
     type = 'dog' and
     "ownerId" in (select "ownerId" from users where type = 'musician') */
const unrelateMusiciansFromDogs = musiciansDogs.unrelate();
```

Some upsides here are that a new class of relation queries can be made with objection, outside the context of a single model instance (i.e. using `$relatedQuery()`).  This makes the API more symmetrical between `Model.query() <--> model.$query()` and `Model.relatedQuery() <--> model.$relatedQuery()`.  Another upside is that `relate()` and `unrelate()` queries can be written without having to build a model instance.

Continuing down this path, here are some additional cases that _may_ be worth supporting,
```js
// per(query)
await User.relatedQuery('pets').per(User.query().findById(100)).relate(200);

// per((builder) => query)
await User.relatedQuery('pets').per((builder) => builder.findById(100)).relate(200);

// per(modifier)
await User.relatedQuery('pets').per('musicians').relate(200);

// per(id) (may be compound)
await User.relatedQuery('pets').per(100).relate(200);

// per(ids) (may be compound)
await User.relatedQuery('pets').per([100, 150]).relate(200);
```

Note that `per()` is just the best name I could come-up with for the purposes of writing-up this issue, but I expect there's possibly/probably a better name out there.

**Comment by koskimas** [comment]
Thanks @devinivy! I'll definitely implement this at some point. I think this is very objectiony feature.

Could `per` be `for` instead? The code would read "related query for 100 where foo = bar"?

**Comment by devinivy** [comment]
Sure thing.  That's exciting!  I'm not attached to `per` at all.  I actually wrote `for` down but decided against it since it's a reserved word in some contexts.  It may be a non-issue, though, if it's always going to be called as a method.

**Comment by koskimas** [comment]
@devinivy I started implementing this it already mostly works. I wanted to do something fun between the tedious 2.0 documentation/renaming stuff :smile: 

Couple of pain points though:

```js
User.relatedQuery('manyToManyRelation').for(subquery).relate(1)
User.relatedQuery('manyToManyRelation').for(subquery).insert(obj)
```

won't work because both of these create inserts. It could work if we just assume the subquery returns a single item, but then the error will be mysterious.

Also multiple ids

```js
User.relatedQuery('manyToManyRelation').for([1, 2]).relate(1)
```
will only work on postgres. There may be a way to insert multiple items using a subquery on some databases, but I don't want to start writing too much db specific code in objection.

**Comment by koskimas** [comment]
Also all these will be impossible for `HasManyRelation`

```js
// per(query)
await User.relatedQuery('pets').per(User.query().findById(100)).relate(200);

// per((builder) => query)
await User.relatedQuery('pets').per((builder) => builder.findById(100)).relate(200);

// per(ids) (may be compound)
await User.relatedQuery('pets').per([100, 150]).relate(200);
```

if the subqueries produce more than one id. If there are multiple identifiers or a subquery that yields multiple values, we would need to create multiple update queries. With one id, it will work:

```js
await User.relatedQuery('pets').per(100).relate(200);
```

```sql
update animals set "ownerId" = 100 where id = 200
```

but with two ids or a subquery, that doesn't make sense.

**Comment by koskimas** [comment]
I'm so excited about this feature! This is exactly what objection is about! Once this is done, I'm going to rewrite a bunch of the old `$relatedQuery` examples to use this instead.

Could this be more difficult to grasp though? People coming from other ORMs are used to this kind of code:

```js
const user = await User.find(100);
const pet = await Animal.find(200);
user.pets.push(pet);
await user.save();
```

There's a pretty long mental leap from that to this:

```js
await User.relatedQuery('pets').for(100).relate(200);
```

What do you think?

**Comment by mordof** [comment]
This looks like `Model.relatedQuery` will return a fairly different QueryBuilder compared to the normal calls.  Will `$relatedQuery` also be updated to use this different QueryBuilder?

Is there room for examining this in a bit of a different perspective, where some specific needs can be identified and added to the default query builder? Maybe it's just me, but this feels like it'd be a bit of a confusing shift using one query builder for most stuff, and then jumping to a different style of api naming/query builder for this `relatedQuery`

**Comment by mordof** [comment]
Nevermind... this is less confusing than I originally thought (primary reason for my confusion is I didn't realize `relate` already exists on `$relatedQuery`). This is just one step further kind of linking  `.relatedQuery` and `$relatedQuery` using `for` - didn't realize that.

**Comment by koskimas** [comment]
@mordof Yep, the only thing that needs to be added is the `for` method. And it can just be added to the normal `QueryBuilder`. The types don't need to take care of only allowing that method after `relatedQuery`. That would be too difficult. Also it really does exist on all query builders anyway. It just does nothing without `relatedQuery`. I could add a runtime error if it's used without it.

**Comment by koskimas** [comment]
First version of this is now implemented in v2.0 branch. That was more work than I thought :sweat_smile:
